Refactor and add support classes for KDBX 4

angelsl opened the pull request:
<!--- Provide a general summary of your changes in the title above -->
The remaining changes are [here](https://github.com/angelsl/keepassxc/tree/kdbx42). I will open a PR once this gets merged.

## Description
<!--- Describe your changes in detail -->
Refer to commit log

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
#148

## How has this been tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? -->
<!--- Please remove all lines which don't apply. -->
- ✅ New feature (non-breaking change which adds functionality)

## Checklist:
<!--- Please go over all the following points. -->
<!--- Again, remove any lines which don't apply. -->
<!--- Pull Requests that don't fulfill all [REQUIRED] requisites are likely -->
<!--- to be sent back to you for correction or will be rejected.  -->
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**
- ✅ My change requires a change to the documentation and I have updated it accordingly.
- ✅ I have added tests to cover my changes.

angelsl commented:
Travis is failing due to too old libgcrypt.

Antidote1911 commented:
nice work !

droidmonkey commented:
Since you renamed the old KeePass2 format read/write to Kdbx3... would it make sense to make the new one Kdbx4.... 

Really we should be instantiating a single common read/write shell class and passing the specific reader/writer as a constructor parameter. This would ensure adding a new format would be painless and independent.

TheZ3ro commented:
@droidmonkey the Kdbx4 classes are in the next [PR](https://github.com/angelsl/keepassxc/commit/f02cd920763ea36729702996a178c27c1cf0c2ce) this is only for the initial renaming

droidmonkey commented:
sweet! Please disregard my post 😄

louib commented on src/core/Database.cpp:
```diff
@@ -210,36 +201,18 @@ void Database::setCompressionAlgo(Database::CompressionAlgorithm algo)
-            }
-        }
+    if (updateTransformSalt) {
+        m_data.kdf->randomizeSalt();
```
@angelsl We could name this function `updateTranformSalt`, or rename the option `randomizeSalt`, just to be consistent.

droidmonkey replied:
randomizeTransformSalt()

angelsl replied:
Done.

louib commented on src/crypto/CryptoHash.cpp:
```diff
@@ -29,27 +29,42 @@ class CryptoHashPrivate
+        qWarning("Gcrypt error (ctor): %s", gcry_strerror(error));
+        qWarning("Gcrypt error (ctor): %s", gcry_strsource(error));
+    }
     Q_ASSERT(error == 0); // TODO: error handling
```
@angelsl not sure how we should be handling those errors, but I guess we should at some point. Right now it'll just fail silently unless the application was built in debug mode.

angelsl replied:
Yeah, but I think this PR is starting to be quite large..

louib commented on src/crypto/CryptoHash.cpp:
```diff
@@ -29,27 +29,42 @@ class CryptoHashPrivate
 
-    int algoGcrypt;
+    int algoGcrypt = -1;
+    unsigned int flagsGcrypt = 0;
```
@angelsl What about using the `GCRY_MD_FLAG_SECURE` flag by default? From the doc:
> Allocate all buffers and the resulting digest in "secure memory". Use this is the hashed data is highly confidential.

angelsl replied:
Added that.

louib commented on src/format/KeePass2RandomStream.cpp:
```diff
@@ -20,16 +20,27 @@
-    return m_cipher.init(CryptoHash::hash(key, CryptoHash::Sha256),
-                         KeePass2::INNER_STREAM_SALSA20_IV);
+    switch (m_cipher.algorithm()) {
+        case SymmetricCipher::Salsa20:
```
@angelsl case and switch on same indentation levels

louib commented on src/format/KeePass2.cpp:
```diff
@@ -0,0 +1,86 @@
+KeePass2::ProtectedStreamAlgo KeePass2::idToProtectedStreamAlgo(quint32 id)
+{
+    switch (id) {
+        case static_cast<quint32>(KeePass2::ArcFourVariant):
```
@angelsl case and switch on same indentation level

droidmonkey commented on src/streams/HashedBlockStream.cpp:
```diff
@@ -256,3 +256,7 @@ bool HashedBlockStream::writeHashedBlock()
     return true;
 }
+
+bool HashedBlockStream::atEnd() const {
```
Rename to isComplete() or isFinished() something like that.

angelsl replied:
This overrides `QIODevice::atEnd`, which *should have been done from the start*

droidmonkey replied:
Ah got it, disregard than

droidmonkey commented on tests/TestKdbx3XmlReader.cpp:
```diff
@@ -0,0 +1,22 @@
+
+#include "TestKeePass2XmlReader.h"
+
+QTEST_GUILESS_MAIN(TestKdbx3XmlReader)
```
Where did the rest of these tests go?

angelsl replied:
Look at CMakeLists.

droidmonkey replied:
OK... so there are two cpp files, one is basically blank. Why are the TestKdbx3XmlReader member functions defined in TestKeePass2XmlReader.cpp??

angelsl replied:
I guess you didn't take a look at the 4 commits that were left out of this PR.

`TestKeePass2XmlReader` has tests that instantiate `KeePass2XmlReader` directly like [this one](https://github.com/angelsl/keepassxc/blob/master/tests/TestKeePass2XmlReader.cpp#L415). In order to avoid having to duplicate the test functions between `Kdbx3XmlReader` and `Kdbx4XmlReader`, I pulled out the XML read and writes into [virtual functions](https://github.com/angelsl/keepassxc/blob/kdbx42/tests/TestKeePass2XmlReader.h#L50) that [use the correct `XmlReader` class](https://github.com/angelsl/keepassxc/blob/kdbx42/tests/TestKeePass2XmlReader.cpp#L81). I decided to keep them in the same source file so that if these functions are ever changed, it is less likely that one half is forgotten, and so that it is easier to verify that they are identical aside from the different `XmlReader` class used.

The better solution, arguably, would have been to use templates, but Qt `moc` does not support templates.

Another solution would have been making the two `XmlReader` classes a subclass of a base `XmlReader` class like the other `Reader` classes, but since the `XmlReader`s are only ever used by the `Reader` classes themselves, I didn't think that was appropriate: the `XmlReader` classes are implementation details of the `Reader` classes and it does not make sense for them to be sibling classes. The only way in which they are similar is that they both take XML data (that *happens* to be similarly structured) and spit out a database. Arguably the `XmlReader` classes should be nested classes of the `Reader` classes, but I digress.

Finally, in case you are wondering why I didn't stick both `QTEST_GUILESS_MAIN` in the `TestKeePass2XmlReader` file, `QTEST_GUILESS_MAIN` is a macro that expands to a definition of `int main`.

droidmonkey commented on tests/TestKeePass2XmlReader.cpp:
```diff
@@ -512,6 +552,10 @@ void TestKeePass2XmlReader::testRepairUuidHistoryItem()
     QVERIFY(!historyItem->uuid().isNull());
     QCOMPARE(historyItem->uuid(), entry->uuid());
+
+    if (db) {
```
You can just ```delete db``` even if its nullptr (its a noop)

angelsl commented on src/format/Kdbx3Reader.cpp:
```diff
@@ -0,0 +1,394 @@
+}
+
+void Kdbx3Reader::setTransformSeed(const QByteArray& data)
+{
```
KDBX 3 doesn't support non-AES KDF at all. The KDF on the database will either be unset or be `AesKdf`. Perhaps the KDF should be set earlier before the header is even read.

droidmonkey replied:
These decisions should be made and enacted by the database class.

angelsl commented on src/format/Kdbx3Reader.cpp:
```diff
@@ -0,0 +1,394 @@
+}
+
+void Kdbx3Reader::setTransformRounds(const QByteArray& data)
+{
```
Same as previous.

angelsl commented on src/format/Kdbx3Reader.h:
```diff
@@ -0,0 +1,62 @@
+public:
+    Kdbx3Reader();
+
+    using BaseKeePass2Reader::readDatabase;
```
This is needed because `readDatabase` is overloaded and the other overload is not virtual.

If this is omitted, the other overload will be hidden.

droidmonkey replied:
That is not true, they are both declared virtual and have different signatures there wouldn't be any shadowing: https://github.com/keepassxreboot/keepassxc/pull/1179/files/f3d019edd927984f249fb57caf0ca35903f5a4b4#diff-3e756dd62a4ab53a029aec11b703b700R38

phoerious replied:
Problem is that C++ is a little wonky when it comes to overloads and overrides. When you override a method it will shadow all overloads of the parent class. So this is indeed needed here.

angelsl commented:
The point of the KDF fields thing is to avoid having these details hardcoded in the GUI.

If you feel that is OK, then yes, we can drop the whole list-of-fields thing, but I will not personally author that.

angelsl commented:
I will work on the rest of the comments in a few days.

It would do you good to look at the rest of the PR, @droidmonkey, so you can understand why some things were done the way they were done, so that I don't have to explain myself repeatedly. We went through some of this in March already, in case you don't remember.

Some of the inconsistencies are there because the original file had them. I'm trying to minimise code churn in this PR. Perhaps you'd like to fix the formatting and whatnot first in a separate PR, and then I, or someone else, can rebase this again.

I'm not very motivated to get this merged, especially if I keep getting comments that result from not fully understanding the codebase. Just laying it out there.

phoerious commented:
> It would do you good to look at the rest of the PR, @droidmonkey, so you can understand why some things were done the way they were done, so that I don't have to explain myself repeatedly. We went through some of this in March already, in case you don't remember.

TBH, things that don't become clear from the code itself, but only from looking at previous discussion about that code need to be documented. Although I know that you will refuse to do that, I would generally prefer doc blocks above every function. I also know that this is a requirement the rest of our code doesn't meet (and therefore I won't refuse to merge because of this), but we want to change that.

droidmonkey commented:
I spent a solid 2.5 days diving into your contribution. If a comment is invalid or I'm being a jerk address it in that specific comment. I'm not unreasonable, this contrib is basically open heart surgery on this app. If this does not work flawlessly, the app is worthless.

droidmonkey commented:
FYI - I deleted the atEnd() comments

hifi commented:
I suggest splitting this even more by leaving all KDBX format related changes out for now (including renames) so the things that everyone agree on can be merged in sooner making future work easier to review and comment on.

I have added a custom KDF in C style from OpenSSH in my PR and it would be really nice to refactor it into a class API that could be defined by this PR. I also used QCryptographicHash for SHA-512 and again this PR could lay the work to bring SHA-512 into the main crypto classes and I could switch to use that instead.

In my eyes this is great work towards the bigger picture and getting even half of the blood and sweat that has gone into it merged will help advance that. Keep up the good work!

phoerious commented on src/gui/DatabaseSettingsWidget.cpp:
```diff
@@ -149,3 +207,66 @@ void DatabaseSettingsWidget::truncateHistories()
+
+void DatabaseSettingsWidget::clearKdfWidgets()
+{
+    m_benchmarkField = nullptr;
```
Drop this function together with the one above.

phoerious commented on src/streams/HmacBlockStream.cpp:
```diff
@@ -0,0 +1,267 @@
+    if (m_error) {
+        return -1;
+    }
+    else if (m_eof) {
```
Remember to not insert newlines before else.

phoerious commented on src/streams/HmacBlockStream.cpp:
```diff
@@ -0,0 +1,267 @@
+
+        m_buffer.clear();
+    }
+    m_blockIndex++;
```
++m_blockIndex;

phoerious commented on src/crypto/kdf/AesKdf.h:
```diff
@@ -0,0 +1,62 @@
+    QByteArray seed() const;
+
+    bool setRounds(quint64 rounds);
+    bool setSeed(const QByteArray& seed);
```
These four methods should be move to the parent class as pure virtual methods. They are generic enough to be applicable to any kind of KDF.

droidmonkey replied:
This would remove the need for fields.

droidmonkey commented on src/core/Database.h:
```diff
@@ -120,6 +115,7 @@ class Database : public QObject
      * Returns a unique id that is only valid as long as the Database exists.
      */
     Uuid uuid();
+    bool changeKdf(Kdf* kdf);
```
These functions probably shouldn't exist at all. The database class should create and maintain the kdf object with external entities merely telling the database which kdf to use.

phoerious commented:
I merged your work into a working branch in our repository. You are very welcome to continue working on this with us, but overall, we'll take it from here. Thank you very much for your contribution!

angelsl commented on src/crypto/kdf/AesKdf.cpp:
```diff
@@ -0,0 +1,171 @@
+    QFuture<bool> future = QtConcurrent::run(transformKeyRaw, raw.left(16), m_seed, m_rounds, &resultLeft);
+
+    bool rightResult = transformKeyRaw(raw.right(16), m_seed, m_rounds, &resultRight);
+    bool leftResult = future.result();
```
This was the original code, FWIW. I only moved it to another class.

angelsl commented:
I haven't had time to look at this until now. Sorry about that.

@phoerious

> TBH, things that don't become clear from the code itself, but only from looking at previous discussion about that code need to be documented.

My comment about some things having been discussed back in March was about certain choices I made, which (I think) would have been clear had this PR not been chopped in half.

> Although I know that you will refuse to do that, I would generally prefer doc blocks above every function.

Indeed, I do not generally think there is a need to document _every single internal function_, but I would've agreed to add that.

@droidmonkey 

> I spent a solid 2.5 days diving into your contribution. If a comment is invalid or I'm being a jerk address it in that specific comment.

I did.

> I'm not unreasonable,

I got pretty frustrated with these comments

>> This function drives me bonkers. Needs to be implemented as a form and when fields get refactored... simplified immensely.
>
>> What is the purpose of the fields if you just store it in a member variable anyway!, I am so confused. The whole kdf code needs to be completely redone and SIMPLIFIED.

because you appear to completely dismiss the fields abstraction without appearing to consider at all why I did it that way.

My original intention was to support arbitrary KDFs, if more are ever added, without having to modify the GUI code.

It is clear both you and @phoerious don't think that is necessary. So be it.

> this contrib is basically open heart surgery on this app. If this does not work flawlessly, the app is worthless and potentially thousands of peoples databases can be corrupted or have flawed security.

I made an effort to make sure (almost?) every single commit in this PR compiles and passes the full test suite. So there's that.

In any case, I *honestly* don't feel like these changes are terribly major. A lot of it was copy and paste, or adding glue code to the crypto classes to support ciphers already supported by gcrypt, etc. A visual review plus manual testing plus the unit test suite should uncover any bugs &mdash; or perhaps I'm being too optimistic.

Perhaps it is time you reformat the entire codebase once through, so that everything is consistent and you can spend less time commenting on stylistic issues. The reason I did weird things like newline-before-`else` was because that was the style used by the source file, and I did not want to go in and reformat everything (too much churn), nor did I want to create a source file with a mix of different styles.

I guess it is for the better that you guys take over the PR. Be sure to pick up the [rest of the KDBX 4 changes](https://github.com/angelsl/keepassxc/tree/kdbx42) as well.

Feel free to contact me if you need any clarifications or help.

phoerious commented on src/crypto/SymmetricCipher.cpp:
```diff
@@ -88,18 +90,57 @@ SymmetricCipher::Algorithm SymmetricCipher::cipherToAlgorithm(Uuid cipher)
     }
+
+    qWarning("SymmetricCipher::cipherToAlgorithm: invalid Uuid %s", cipher.toByteArray().toHex().data());
+    return InvalidAlgorithm;
```
cipher is of type Uuid, not int.

sebast889 commented:
>I guess it is for the better that you guys take over the PR. Be sure to pick up the rest of the KDBX 4 changes as well.

@phoerious were these changes ever picked up? If not, are they are still needed?

droidmonkey commented:
This PR is merged.....

sebast889 commented:
I realize that. I just thought those changes in the linked repo might have been forgotten about. If they weren't then disregard my comment.

droidmonkey commented:
Yes we completely overhauled this implementation. It's been in since 2.3.0
